### PR TITLE
!!! FEATURE: Extend the QueueInterface with countReserved and countFailed 

### PR DESCRIPTION
**Pull request by daniellienert**
Also add countReady() and deprecate the count() command
to clearify the interface

### CODE REVIEW

File: Classes/Command/QueueCommandController.php [hunk 1]
```diff
@@ -40,14 +42,15 @@ class QueueCommandController extends CommandController
      * Displays all configured queues, their type and the number of messages that are ready to be processed.
      *
      * @return void
+     * @throws Exception
```
Comment (bwaidelich) [inline Classes/Command/QueueCommandController.php]: Our CGL don't state anything about the requirement of `@throws` annotations. Personally I'm strongly [against those](https://discuss.neos.io/t/to-throws-or-not-to-throws-indirect-exceptions/3018/4).
Obviously no reason to block this, but if we decide to add the annotations we should do it for the whole package, and maybe in a separate PR

File: Classes/Command/QueueCommandController.php [hunk 2]
```diff
@@ -11,9 +11,11 @@
  * source code.
  */
 
+use Flowpack\JobQueue\Common\Exception;
 use Flowpack\JobQueue\Common\Queue\QueueManager;
 use Neos\Flow\Annotations as Flow;
 use Neos\Flow\Cli\CommandController;
+use Neos\Flow\Mvc\Exception\StopActionException;
```
Comment (bwaidelich) [inline Classes/Command/QueueCommandController.php]: Why is this namespace imported?

File: Classes/Queue/QueueInterface.php [hunk 3]
```diff
@@ -102,13 +102,33 @@ public function finish($messageId);
...
 
     /**
-     * Count ready messages in the queue
+     * Get a count of ready messages currently in the queue.
      *
-     * Get a count of messages currently in the queue.
+     * @return int The number of ready messages in the queue
+     * @deprecated Will be removed with the next major. Use countReady instead.
+     */
+    public function count(): int;
```
Comment (bwaidelich) [inline Classes/Queue/QueueInterface.php]: If we change the interface we don't need to keep deprecated methods IMO as the implementations need to be adjusted anyways

File: Tests/Functional/AbstractQueueTest.php [hunk 4]
```diff
@@ -179,16 +179,32 @@ public function abortRemovesMessageFromActiveQueue()
...
+        $this->assertSame(0, $this->queue->countReady());
+    }
+
+    /**
+     * @test
+     */
+    public function countFailedReturnsZeroByDefault()
+    {
+        $this->assertSame(0, $this->queue->countFailed());
+    }
```
Comment (bwaidelich) [inline Tests/Functional/AbstractQueueTest.php]: We should add tests effectively covering the new features:
`countFailedReturnsNumberOfFailedMessages()` and `countReservedReturnsNumberOfReservedMessages()`

File: Classes/Command/QueueCommandController.php (hunk 2 above)
Comment (bwaidelich) [reply Classes/Command/QueueCommandController.php]: Nevermind, it's needed for the `@throws` annotations below

File: Tests/Functional/AbstractQueueTest.php (hunk 4 above)
Comment (bwaidelich) [reply Tests/Functional/AbstractQueueTest.php]: Added with bdffb4a5cef61544aeaffd999d85a9c6db799190

**Comment by daniellienert** [comment]
@bwaidelich: Thanks a lot! 
For the `@throws`: lets find a common approach in discuss and adapt the code if needed.
For the return types: I think it would be good to change the complete interface to use return types, as it is also breaking, before releasing a new major. But didn't want to make this PR too big.

**Pull request merged**